blockchain/internal/spendpruner/db.go

```diff
@@ -76,10 +85,77 @@ func dbUpdateSpendConsumerDeps(dbTx database.Tx, blockHash chainhash.Hash, consu
 	return depsBucket.Put(blockHash[:], serialized)
 }
 
-// dbFetchSpendConsumerDeps uses an existing database transaction to fetch all
-// spend consumer dependency entries in the database.
-func dbFetchSpendConsumerDeps(dbTx database.Tx) (map[chainhash.Hash][]string, error) {
-	depsBucket := dbTx.Metadata().Bucket(spendConsumerDepsBucketName)
+// dbPersistSpendHeights uses an existing database transaction to persist the
+// provided spend heights.
+func dbPersistSpendHeights(dbTx database.Tx, spendHeights map[chainhash.Hash]uint32) error {
+	heightsBucket := dbTx.Metadata().Bucket(spendJournalHeightsBucketName)
+
+	// return immediately if there are no spend heights to persist.
+	if len(spendHeights) == 0 {
+		return nil
+	}
+
+	// Persist all spend height map entries.
+	for blockHash, height := range spendHeights {
```

Comment:
All of the code in this commit is chalk full of incorrect usage.  The `range` statement overwrites the variables with each iteration and you're slicing them directly and passing them on to db code, so with any type of async behavior, it will be a race and overwrite things incorrectly.

```Go
package main

import (
	"encoding/binary"
	"fmt"
	"time"
)

func put(k []byte, v []byte) {
	go func(k, v []byte) {
		time.Sleep(time.Second)
		fmt.Printf("k: %x, v: %x\n", k, v)
	}(k, v)
}

func main() {
	foo := map[[32]byte]uint32{
		[32]byte{0x00}: 0,
		[32]byte{0x01}: 1,
		[32]byte{0x02}: 2,
	}

	for hash, height := range foo {
		var b [4]byte
		binary.LittleEndian.PutUint32(b[:], height)
		put(hash[:], b[:])
	}

	time.Sleep(time.Second * 2)
}
```
Output:
```
k: 0200000000000000000000000000000000000000000000000000000000000000, v: 00000000
k: 0200000000000000000000000000000000000000000000000000000000000000, v: 01000000
k: 0200000000000000000000000000000000000000000000000000000000000000, v: 02000000
```

Reply:
Noted, thanks. Updating.